### ISSUE TITLE
Nagios Integration / Nagios plugin for Cachet

### ISSUE THREAD

**Issue by sammcj** [cachethq/Cachet#225]
It would be neat to have a Nagios plugin that updated Cachet.

I think this would be pretty easy via the API but I'd recommend having a separate user for it (https://github.com/cachethq/Cachet/issues/128).

I'll keep an eye on https://github.com/cachethq/Cachet/issues/128 then I might whip something up for this.

**Comment by GrahamCampbell** [comment]
:+1:

**Comment by MichaelBanks** [comment]
:+1: but, surely this'd just be some config on the Nagios server side. You could just create a contact for Cachet and get it to call the API when something's going on, similar to Trello's integration with Nagios.

**Comment by jbrooksuk** [comment]
@MichaelBanks did you get any further on this integration?

**Comment by MichaelBanks** [comment]
@jbrooksuk I think we were going to have a chat about it at some point. Give me a ring when you get a chance

**Comment by jbrooksuk** [comment]
:+1:

Once my voice is better, I'll do that :)

**Comment by sammcj** [comment]
Hey, I think we'll be able to write a decent Nagios check but it won't be until after multi-app is at least in beta - (after 1.0 release) - We've got a guy who's great at writing reusable Nagios checks, so if you wanted to put this on the burner until later we (@infoxchange) can do that.

**Comment by jbrooksuk** [comment]
Amazing, thanks Sam.

Once V1 is out, we'll come back to this.

**Comment by sammcj** [comment]
@georgehansper I'll speak to you about this one since you're a monitoring guru ;)

**Comment by MichaelBanks** [comment]
I haven't been keeping up with Cachet development as much as I'd like. But last time I checked, we came to a fairly solid conclusion that Cachet would need some kind of configurable hooks setup. A more fluid API layer. Nagios has some set environment variables which are passed through when alerts on hosts and services pop up, so Cachet has to seat those variables, otherwise we have to go through and truncate everything down in a perl script on the server side which gets far too complicated.

**Comment by MichaelBanks** [comment]
Here's the script Slack uses for Nagios integration: https://github.com/tinyspeck/services-examples/blob/master/nagios.pl

**Comment by MichaelBanks** [comment]
This is the command Nagios would pass when a service notifcation occurs:

`/usr/local/bin/slack_nagios.pl -field slack_channel=#alerts -field HOSTALIAS="$HOSTNAME$" -field SERVICEDESC="$SERVICEDESC$" -field SERVICESTATE="$SERVICESTATE$" -field SERVICEOUTPUT="$SERVICEOUTPUT$" -field NOTIFICATIONTYPE="$NOTIFICATIONTYPE$"`
and a host notification:
`/usr/local/bin/slack_nagios.pl -field slack_channel=#ops -field HOSTALIAS="$HOSTNAME$" -field HOSTSTATE="$HOSTSTATE$" -field HOSTOUTPUT="$HOSTOUTPUT$" -field NOTIFICATIONTYPE="$NOTIFICATIONTYPE$"`

**Comment by MichaelBanks** [comment]
The resulting post request would be as follows (for service notifications):
```
https://<slack domain>/services/hooks/nagios?token=<token_from_slack>
{
"HOSTALIAS":"$HOSTNAME$"
"SERVICEDESC":"$SERVICEDESC$"
"SERVICEOUTPUT":"$SERVICEOUTPUT$"
... etc
}
```

**Comment by sammcj** [comment]
@jbrooksuk ,

I've just given @georgehansper from our systems engineering team @infoxchange two hours on writing an initial Nagios check for Cachet - do you have a running instance of cachet he can test the check against?

**Comment by georgehansper** [comment]
A Nagios check for this should look at /api/components

Each component should be examined in turn, and an overall status returned based on the 'worst case'

Looking at the status codes in app/lang/en/cachet.php, I think the status code should map to Nagios codes as follows:
1= Operational - > Nagios OK
2 = Performance Issues -> Nagios Warning
3 = Partial Outage -> Nagios Critical
4 = Major Outage -> Nagios Critical

I'm not sure how the 'incident_count' could or should be used.
Is this a count of past incidents?
If so, it could be used by Nagios as a performance metric.

It's also possible to examine the/api/incidents but if this is more of a hsitorical list of previous incidents, it probably not suitable for a Nagios check.

**Comment by jbrooksuk** [comment]
You're more than welcome to use the demo; https://status.cachethq.io:

- test@test.com
- test123

@georgehansper there isn't an incident count actually, but that would be useful.

**Comment by georgehansper** [comment]
I've written a Nagios check that uses /api/components, and that works fine.

But I think the more valuable side would be to get Nagios to set the status of a component.

I've tried using the API, both with the perl JSON library, and using curl, and I can't seem to get it to work.

The curl command I'm using is:
curl -u 'test@test.com:test123' -H "Content-Type: application/json" -d '{"name":"API 2","status":1}' -H 'User-Agent: Mozilla/5.0 (but really curl)' https://status.cachethq.io/api/components -v

The result is:
HTTP/1.1 500 Internal Server Error

There is a message on the error page:
SQLSTATE[23502]: Not null violation: 7 ERROR: null value in column "description" violates not-null constraint DETAIL: Failing row contains (15, API 2, null, null, 1, 0, 0, 1, 2015-02-19 22:58:40, 2015-02-19 22:58:40, null). (SQL: insert into "components" ("order", "group_id", "name", "status", "user_id", "updated_at", "created_at") values (0, 0, API 2, 1, 1, 2015-02-19 22:58:40, 2015-02-19 22:58:40) returning "id”)

I think the API is assuming I want to _create_ a component, when actually, I just want to update it.

The User-Agent header is there because I know CloudFlare would not let me use the perl LWP user agent until I change this header.

**Comment by jbrooksuk** [comment]
Thanks, can you open that as a separate issue please? It's the cross-database issue again. CC @cachethq/owners.

**Comment by GrahamCampbell** [comment]
Grrr. I assume you're using posgres? Mysql is so dumb it means we're missing all these violations.

**Comment by sammcj** [comment]
I thought George was just testing against the public demo page - @jbrooksuk I'm assuming you still use MySQL with that?

When we start to deploy Cachet internally - we'll definitely be using PostgreSQL if we're able to - we've just had too many issues with MySQL over the years and have finally rid ourselves of the last MySQL instance last year, and that wasn't very fun - the quality of the data in large MySQL databases becomes very inconsistent / dirty over time.

**Comment by jbrooksuk** [comment]
I always use MySQL, which means when we deploy to PgSQL it breaks. For everything.

**Comment by anarcat** [comment]
:+1:

**Comment by jbrooksuk** [comment]
@georgehansper did you get anywhere with your Nagios integration? I'd love to share it as an integration.

**Comment by sammcj** [comment]
I think @georgehansper was waiting on updated API documentation or something to that effect that's probably resolved now anyway :) He's just working on some virtualisation upgrades for us today but I'm sure we can spare him for a few hours later this week?

**Comment by jbrooksuk** [comment]
Maybe next week then? I'm updating our demo to the latest version on Friday morning 6:30AM GMT+0 so it'll be easier to stay consistent.

**Comment by sammcj** [comment]
Sounds like a plan, I'll ask George lock out some of his calendar next week :) GL with the update James!

**Comment by jbrooksuk** [comment]
This isn't something we'll be supporting natively from Cachet but hopefully @georgehansper and @sammcj can sort us out with something :)
